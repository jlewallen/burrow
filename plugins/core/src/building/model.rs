use crate::library::model::*;

use std::str::FromStr;

#[derive(Clone, Serialize, Deserialize, PartialEq, ToTaggedJson, Reply, Debug)]
#[serde(rename_all = "camelCase")]
pub enum Constructed {
    Area(AreaObservation),
}

impl TryFrom<Constructed> for Effect {
    type Error = TaggedJsonError;

    fn try_from(value: Constructed) -> std::result::Result<Self, Self::Error> {
        Ok(Self::Reply(value.to_tagged_json()?.into()))
    }
}

#[derive(Debug, Serialize, ToTaggedJson)]
#[serde(rename_all = "camelCase")]
struct EditorReply {}

impl Reply for EditorReply {}

#[derive(Default, Clone, Debug)]
pub struct QuickEdit {
    pub name: Option<String>,
    pub desc: Option<String>,
}

impl TryFrom<&EntityPtr> for QuickEdit {
    type Error = DomainError;

    fn try_from(value: &EntityPtr) -> std::result::Result<Self, Self::Error> {
        let name = Some(value.name()?);
        let desc = value.desc()?;
        Ok(Self { name, desc })
    }
}

const SEPARATOR: &str =
    "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~";

impl FromStr for QuickEdit {
    type Err = DomainError;

    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
        let parts: Vec<_> = s.split(SEPARATOR).collect();
        if parts.len() != 2 {
            return Err(DomainError::Anyhow(anyhow::anyhow!("malformed quick edit")));
        }

        let (name, desc) = match parts[..] {
            [name, desc] => (name, desc),
            _ => todo!(),
        };

        let name = Some(name.trim().to_owned());
        let desc = Some(desc.trim().to_owned());

        Ok(Self { name, desc })
    }
}

impl ToString for QuickEdit {
    fn to_string(&self) -> String {
        format!(
            "{}\n\n{}\n\n{}",
            self.name.as_deref().unwrap_or(""),
            SEPARATOR,
            self.desc.as_deref().unwrap_or("")
        )
    }
}
