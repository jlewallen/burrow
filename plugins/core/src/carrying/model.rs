use crate::{library::model::*, tools};

pub use kernel::common::Carrying;

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct Containing {
    pub(crate) holding: Vec<EntityRef>,
    pub(crate) capacity: Option<u32>,
    pub(crate) produces: HashMap<String, String>,
}

impl Scope for Containing {
    fn scope_key() -> &'static str {
        "containing"
    }
}

impl Containing {
    pub fn start_carrying(&mut self, item: &EntityPtr) -> Result<bool, DomainError> {
        if let Some(carryable) = item.scope::<Carryable>()? {
            let holding = self
                .holding
                .iter()
                .map(|h| h.to_entity())
                .collect::<Result<Vec<_>, _>>()?;

            for held in holding {
                if is_kind(&held, &carryable.kind)? {
                    let mut combining = held.scope_mut::<Carryable>()?;

                    combining.increase_quantity(&carryable.quantity.into())?;

                    combining.save()?;

                    get_my_session()?.obliterate(item)?;

                    return Ok(true);
                }
            }
        }

        if !self.is_holding(item) {
            self.holding.push(item.entity_ref());
        }

        Ok(true)
    }

    pub fn is_holding(&self, item: &EntityPtr) -> bool {
        self.holding.iter().any(|i| *i.key() == item.key())
    }

    fn remove_item(&mut self, item: &EntityPtr) -> Result<bool, DomainError> {
        self.holding = self
            .holding
            .iter()
            .flat_map(|i| {
                if *i.key() == item.key() {
                    vec![]
                } else {
                    vec![i.clone()]
                }
            })
            .collect::<Vec<EntityRef>>()
            .to_vec();

        Ok(true)
    }

    pub fn stop_carrying(&mut self, found: Found) -> Result<Option<EntityPtr>, DomainError> {
        let item = found.entity()?;

        if !self.is_holding(item) {
            return Ok(None);
        }

        if let Some(carryable) = item.scope::<Carryable>()? {
            match &found {
                Found::One(item) => {
                    if carryable.quantity > 1.0 {
                        let (_original, separated) = tools::separate(item, &1.0.into())?;

                        Ok(Some(separated))
                    } else {
                        self.remove_item(item)?;

                        Ok(Some(item.clone()))
                    }
                }
                Found::Quantified(q, entity) => {
                    if carryable.quantity > q.as_f32() {
                        let (_original, separated) = tools::separate(entity, q)?;

                        Ok(Some(separated))
                    } else {
                        Ok(None)
                    }
                }
            }
        } else {
            self.remove_item(item)?;

            Ok(Some(item.clone()))
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Carryable {
    kind: Kind,
    quantity: f32,
}

fn is_kind(entity: &EntityPtr, kind: &Kind) -> Result<bool, DomainError> {
    if let Some(carryable) = entity.scope::<Carryable>()? {
        Ok(*carryable.kind() == *kind)
    } else {
        Ok(false)
    }
}

impl Default for Carryable {
    fn default() -> Self {
        let session = get_my_session().expect("No session in Entity::new_blank!");
        Self {
            kind: Kind::new(session.new_identity()),
            quantity: 1.0,
        }
    }
}

impl Carryable {
    pub fn quantity(&self) -> f32 {
        self.quantity
    }

    pub fn decrease_quantity(&mut self, q: &Quantity) -> Result<&mut Self, DomainError> {
        self.sanity_check_quantity();

        if *q < 1.0.into() || *q > self.quantity.into() {
            Err(DomainError::Impossible)
        } else {
            self.quantity -= q.as_f32();

            Ok(self)
        }
    }

    pub fn increase_quantity(&mut self, q: &Quantity) -> Result<&mut Self, DomainError> {
        self.sanity_check_quantity();

        self.quantity += q.as_f32();

        Ok(self)
    }

    pub fn set_quantity(&mut self, q: &Quantity) -> Result<&mut Self, DomainError> {
        self.quantity = q.as_f32();

        Ok(self)
    }

    pub fn kind(&self) -> &Kind {
        &self.kind
    }

    pub fn set_kind(&mut self, kind: &Kind) {
        self.kind = kind.clone();
    }

    // Migrate items that were initialized with 0 quantities.
    fn sanity_check_quantity(&mut self) {
        if self.quantity < 1.0 {
            self.quantity = 1.0
        }
    }
}

impl Scope for Carryable {
    fn scope_key() -> &'static str {
        "carryable"
    }
}
