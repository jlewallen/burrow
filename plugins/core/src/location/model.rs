use crate::library::model::*;

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct Location {
    pub container: Option<EntityRef>,
}

impl Location {
    pub fn set(item: &EntityPtr, container: EntityRef) -> Result<(), DomainError> {
        let mut location = item.scope_mut::<Location>()?;
        location.container = Some(container);
        location.save()
    }

    pub fn get(item: &EntityPtr) -> Result<Option<EntityRef>, DomainError> {
        match item.scope::<Location>()? {
            Some(location) => Ok(location.container.clone()),
            None => Ok(None),
        }
    }
}

impl Scope for Location {
    fn scope_key() -> &'static str {
        "location"
    }
}

pub fn change_location<A, B, C, D>(
    from: &EntityPtr,
    to: &EntityPtr,
    item: &EntityPtr,
    do_from: C,
    do_into: D,
) -> Result<bool, DomainError>
where
    A: Scope + Serialize,
    B: Scope + Serialize,
    C: FnOnce(&mut A, EntityPtr) -> Result<Option<EntityPtr>, DomainError>,
    D: FnOnce(&mut B, EntityPtr) -> Result<Option<EntityPtr>, DomainError>,
{
    info!("moving {:?} {:?} {:?}", item, from, to);

    let mut from = from.scope_mut::<A>()?;
    let mut into = to.scope_mut::<B>()?;

    match do_from(&mut from, item.clone())? {
        Some(moving) => match do_into(&mut into, moving)? {
            Some(moving) => {
                Location::set(&moving, to.entity_ref())?;
                from.save()?;
                into.save()?;

                Ok(true)
            }
            None => Ok(false),
        },
        None => Ok(false),
    }
}

pub fn container_of(item: &EntityPtr) -> Result<EntityPtr, DomainError> {
    Location::get(item)?
        .ok_or(DomainError::ContainerRequired)?
        .to_entity()
}
