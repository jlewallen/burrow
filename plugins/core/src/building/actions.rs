use std::{rc::Rc, str::FromStr};

use chrono::Utc;

use crate::{
    building::model::{Constructed, QuickEdit},
    carrying::model::{Carryable, Containing},
    library::actions::*,
    looking::{actions::LookAction, model::new_area_observation},
    memory::model::{remember, EntityEvent, MemoryEvent},
    moving::model::Occupyable,
};

#[action]
pub struct AddScopeAction {
    pub scope_key: String,
}

impl Action for AddScopeAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, _session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        // TODO Security wise we would probably need a global list of who could
        // manipulate which scopes eventually.
        // TODO Right now this requires scopes to be functionable if all their
        // fields are ommitted. Look into `#[serde(default)]` to make this work
        // w/o a bunch of Option's?
        let Some(item) = tools::holding_one_item(surroundings.living())? else {
            return Ok(SimpleReply::NotFound.try_into()?);
        };

        debug!(item = ?item, scope_key = %self.scope_key, "add-scope");

        let mut item = item.entity().borrow_mut();

        item.add_scope_by_key(&self.scope_key);

        Ok(SimpleReply::Done.try_into()?)
    }
}

#[action]
pub struct EditAction {
    pub item: Item,
}

impl Action for EditAction {
    fn is_read_only() -> bool {
        true
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        debug!("editing {:?}!", self.item);

        match session.find_item(surroundings, &self.item)? {
            Some(editing) => {
                info!("editing {:?}", editing);
                let quick_edit: QuickEdit = (&editing).try_into()?;
                Ok(EditorReply::new(
                    editing.key().to_string(),
                    WorkingCopy::Markdown(quick_edit.to_string()),
                    SaveQuickEditAction::new_template(editing.key().clone())?,
                )
                .try_into()?)
            }
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct EditRawAction {
    pub item: Item,
}

impl Action for EditRawAction {
    fn is_read_only() -> bool {
        true
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        debug!("editing {:?}!", self.item);

        match session.find_item(surroundings, &self.item)? {
            Some(editing) => {
                info!("editing {:?}", editing);
                let json = {
                    let editing = editing.borrow();
                    editing.to_json_value()?
                };
                let key = editing.key().clone();
                Ok(EditorReply::new(
                    key.to_string(),
                    WorkingCopy::Json(json),
                    SaveEntityJsonAction::new_template(key)?,
                )
                .try_into()?)
            }
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct DuplicateAction {
    pub item: Item,
}

impl Action for DuplicateAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("duplicating {:?}!", self.item);

        match session.find_item(surroundings, &self.item)? {
            Some(duplicating) => {
                info!("duplicating {:?}", duplicating);
                _ = tools::duplicate(&duplicating)?;
                Ok(SimpleReply::Done.try_into()?)
            }
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct ObliterateAction {
    pub item: Item,
}

impl Action for ObliterateAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("obliterate {:?}!", self.item);

        match session.find_item(surroundings, &self.item)? {
            Some(obliterating) => {
                info!("obliterate {:?}", obliterating);
                tools::obliterate(&obliterating)?;
                Ok(SimpleReply::Done.try_into()?)
            }
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct MakeItemAction {
    pub name: String,
}

impl Action for MakeItemAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("make-item {:?}", self.name);

        let creator = surroundings.living();

        let new_item: Entity = build_entity()
            .default_scope::<Carryable>()?
            .creator(creator.entity_ref())
            .name(&self.name)
            .try_into()?;

        let new_item = session.add_entity(new_item)?;

        tools::set_quantity(&new_item, 1f32)?;
        tools::set_container(creator, &vec![new_item.clone()])?;

        remember(
            &creator,
            Utc::now(),
            MemoryEvent::Created(EntityEvent {
                key: new_item.key().clone(),
                gid: new_item.gid(),
                name: new_item.name()?.unwrap(),
            }),
        )?;

        Ok(SimpleReply::Done.try_into()?)
    }
}

#[action]
pub struct BidirectionalDigAction {
    pub outgoing: String,
    pub returning: String,
    pub new_area: String,
}

impl Action for BidirectionalDigAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!(
            "bidirectional-dig {:?} <-> {:?} '{:?}'",
            self.outgoing, self.returning, self.new_area
        );

        let (_, living, area) = surroundings.unpack();

        let new_area: Entity = build_entity()
            .area()
            .default_scope::<Occupyable>()?
            .default_scope::<Containing>()?
            .name(&self.new_area)
            .desc(&self.new_area)
            .try_into()?;
        let new_area = session.add_entity(new_area)?;

        tools::add_route(&area, &self.outgoing, &new_area)?;
        tools::add_route(&new_area, &self.returning, &area)?;

        match tools::navigate_between(&area, &new_area, &living)? {
            true => Ok(session.perform(Perform::Living {
                living,
                action: PerformAction::Instance(Rc::new(LookAction {})),
            })?),
            false => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[derive(Debug, Serialize, Deserialize, ToTaggedJson, DeserializeTagged)]
pub struct SaveQuickEditAction {
    pub key: EntityKey,
    pub copy: WorkingCopy,
}

impl SaveQuickEditAction {
    pub fn new(key: EntityKey, copy: WorkingCopy) -> Self {
        Self { key, copy }
    }

    pub fn new_template(key: EntityKey) -> Result<JsonTemplate, TaggedJsonError> {
        let copy = WorkingCopy::Markdown(JSON_TEMPLATE_VALUE_SENTINEL.to_owned());
        let template = Self { key, copy };

        Ok(template.to_tagged_json()?.into())
    }
}

impl Action for SaveQuickEditAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, _surroundings: &Surroundings) -> ReplyResult {
        info!("save:quick-edit {:?}", self.key);

        match session.entity(&LookupBy::Key(&self.key))? {
            Some(entity) => match &self.copy {
                WorkingCopy::Markdown(text) => {
                    let quick = QuickEdit::from_str(text)?;
                    let mut entity = entity.borrow_mut();
                    if let Some(name) = quick.name {
                        entity.set_name(&name)?;
                    }
                    if let Some(desc) = quick.desc {
                        entity.set_desc(&desc)?;
                    }

                    Ok(SimpleReply::Done.try_into()?)
                }
                _ => Err(anyhow::anyhow!("Save expected JSON working copy")),
            },
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[derive(Debug, Serialize, Deserialize, ToTaggedJson, DeserializeTagged)]
pub struct SaveEntityJsonAction {
    pub key: EntityKey,
    pub copy: WorkingCopy,
}

impl SaveEntityJsonAction {
    pub fn new(key: EntityKey, copy: WorkingCopy) -> Self {
        Self { key, copy }
    }

    pub fn new_template(key: EntityKey) -> Result<JsonTemplate, TaggedJsonError> {
        let copy = WorkingCopy::Json(JsonValue::String(JSON_TEMPLATE_VALUE_SENTINEL.to_owned()));
        let template = Self { key, copy };

        Ok(template.to_tagged_json()?.into())
    }
}

impl Action for SaveEntityJsonAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, _surroundings: &Surroundings) -> ReplyResult {
        info!("save:entity-json {:?}", self.key);

        match session.entity(&LookupBy::Key(&self.key))? {
            Some(entity) => match &self.copy {
                WorkingCopy::Json(value) => {
                    let replacing = Entity::from_value(value.clone())?;
                    entity.replace(replacing);

                    Ok(SimpleReply::Done.try_into()?)
                }
                _ => Err(anyhow::anyhow!("Save expected JSON working copy")),
            },
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct BuildAreaAction {
    pub name: String,
}

impl Action for BuildAreaAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("build-area {:?}", self.name);

        let creator = surroundings.living();

        let new_area: Entity = build_entity()
            .area()
            .default_scope::<Containing>()?
            .default_scope::<Occupyable>()?
            .creator(creator.entity_ref())
            .name(&self.name)
            .try_into()?;

        let new_area = session.add_entity(new_area)?;

        remember(
            &creator,
            Utc::now(),
            MemoryEvent::Constructed(EntityEvent {
                key: new_area.key().clone(),
                gid: new_area.gid(),
                name: new_area.name()?.unwrap(),
            }),
        )?;

        info!("created {:?}", new_area);

        let observed = new_area_observation(creator, &new_area)?;
        let reply = Constructed::Area(observed);
        Ok(reply.try_into()?)
    }
}
