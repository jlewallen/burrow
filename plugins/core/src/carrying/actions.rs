use crate::{carrying::model::Carrying, library::actions::*, looking::model::Observe};

#[action]
pub struct HoldAction {
    pub item: Item,
}

impl Action for HoldAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("hold {:?}!", self.item);

        let (_, actor, area) = surroundings.unpack();

        match session.find_item(surroundings, &self.item)? {
            Some(holding) => match tools::move_between(&area, &actor, holding.clone())? {
                true => Ok(reply_ok(
                    actor.clone(),
                    Audience::Area(area.key().clone()),
                    Carrying::Held {
                        actor: (&actor).observe(&actor)?.expect("No observed entity"),
                        item: (&holding.entity()?)
                            .observe(&actor)?
                            .expect("No observed entity"),
                        area: (&area).observe(&actor)?.expect("No observed entity"),
                    },
                )?),
                false => Ok(SimpleReply::NotFound.try_into()?),
            },
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct DropAction {
    pub maybe_item: Option<Item>,
}

impl Action for DropAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("drop {:?}!", self.maybe_item);

        let (_, actor, area) = surroundings.unpack();

        match &self.maybe_item {
            Some(item) => match session.find_item(surroundings, item)? {
                Some(dropping) => match tools::move_between(&actor, &area, dropping.clone())? {
                    true => Ok(reply_ok(
                        actor.clone(),
                        Audience::Area(area.key().clone()),
                        Carrying::Dropped {
                            actor: (&actor).observe(&actor)?.expect("No observed entity"),
                            item: (&dropping.entity()?)
                                .observe(&actor)?
                                .expect("No observed entity"),
                            area: (&area).observe(&actor)?.expect("No observed entity"),
                        },
                    )?),
                    false => Ok(SimpleReply::NotFound.try_into()?),
                },
                None => Ok(SimpleReply::NotFound.try_into()?),
            },
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct PutInsideAction {
    pub item: Item,
    pub vessel: Item,
}

impl Action for PutInsideAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("put-inside {:?} -> {:?}", self.item, self.vessel);

        let (_, _user, _area) = surroundings.unpack();

        match session.find_item(surroundings, &self.item)? {
            Some(item) => match session.find_item(surroundings, &self.vessel)? {
                Some(vessel) => {
                    let vessel = vessel.one()?;
                    if tools::is_container(&vessel)? {
                        let from = tools::container_of(&item.clone().one()?)?;
                        match tools::move_between(&from, &vessel, item)? {
                            true => Ok(SimpleReply::Done.try_into()?),
                            false => Ok(SimpleReply::NotFound.try_into()?),
                        }
                    } else {
                        Ok(SimpleReply::Impossible.try_into()?)
                    }
                }
                None => Ok(SimpleReply::NotFound.try_into()?),
            },
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct TakeOutAction {
    pub item: Item,
    pub vessel: Item,
}

impl Action for TakeOutAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("take-out {:?} -> {:?}", self.item, self.vessel);

        let (_, user, _area) = surroundings.unpack();

        match session.find_item(surroundings, &self.vessel)? {
            Some(vessel) => {
                let vessel = vessel.one()?;
                if tools::is_container(&vessel)? {
                    match session.find_item(surroundings, &self.item)? {
                        Some(item) => match tools::move_between(&vessel, &user, item)? {
                            true => Ok(SimpleReply::Done.try_into()?),
                            false => Ok(SimpleReply::NotFound.try_into()?),
                        },
                        None => Ok(SimpleReply::NotFound.try_into()?),
                    }
                } else {
                    Ok(SimpleReply::Impossible.try_into()?)
                }
            }
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct GiveToAction {
    pub item: Item,
    pub receiver: Item,
}

impl Action for GiveToAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("give-to {:?} -> {:?}", self.item, self.receiver);

        let (_, user, _area) = surroundings.unpack();

        // I think there are very interesting permission related implications
        // here. For example, limiting third party access to your hands except
        // for key individuals.
        match session.find_item(surroundings, &self.item)? {
            Some(item) => match session.find_item(surroundings, &self.receiver)? {
                Some(receiver) => match tools::move_between(&user, &receiver.one()?, item)? {
                    true => Ok(SimpleReply::Done.try_into()?),
                    false => Ok(SimpleReply::NotFound.try_into()?),
                },
                None => Ok(SimpleReply::NotFound.try_into()?),
            },
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}
