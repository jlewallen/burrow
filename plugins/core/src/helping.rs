use crate::library::plugin::*;

#[derive(Default)]
pub struct HelpingPluginFactory {}

impl PluginFactory for HelpingPluginFactory {
    fn create_plugin(&self) -> Result<Box<dyn Plugin>> {
        Ok(Box::new(HelpingPlugin {}))
    }

    fn stop(&self) -> Result<()> {
        Ok(())
    }
}

#[derive(Default)]
pub struct HelpingPlugin {}

impl Plugin for HelpingPlugin {
    fn plugin_key() -> &'static str
    where
        Self: Sized,
    {
        "helping"
    }

    fn key(&self) -> &'static str {
        Self::plugin_key()
    }

    fn initialize(&mut self) -> Result<()> {
        Ok(())
    }

    fn middleware(&mut self) -> Result<Vec<Rc<dyn Middleware>>> {
        Ok(Vec::default())
    }

    fn deliver(&self, _incoming: &Incoming) -> Result<()> {
        Ok(())
    }

    fn stop(&self) -> Result<()> {
        Ok(())
    }
}

impl ParsesActions for HelpingPlugin {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        try_parsing(parser::HelpWithParser {}, i)
            .or_else(|_| try_parsing(parser::ReadHelpParser {}, i))
    }
}

pub mod model {
    use crate::library::model::*;

    #[derive(Debug, Serialize, Deserialize, Default)]
    pub struct Page {
        pub body: String,
    }

    #[derive(Debug, Serialize, Deserialize, Default)]
    pub struct Wiki {
        // pub help: Option<HashMap<String, Page>>,
        pub acls: Option<Acls>,
        pub body: Option<String>,
    }

    // const DEFAULT_HELP: &str = "default";

    impl Wiki {
        pub fn get_default(&self) -> Option<String> {
            /*
            self.help
                .as_ref()
                .and_then(|h| h.get(DEFAULT_HELP).map(|h| h.body.to_owned()))
            */
            self.body.clone()
        }

        pub fn set_default(&mut self, value: &str) {
            /*
            let help = self.help.get_or_insert_with(|| HashMap::default());
            help.insert(
                DEFAULT_HELP.to_owned(),
                Page {
                    body: value.to_owned(),
                },
            );
            */
            self.body = Some(value.to_owned());
        }
    }

    impl Scope for Wiki {
        fn serialize(&self) -> Result<serde_json::Value> {
            Ok(serde_json::to_value(self)?)
        }

        fn scope_key() -> &'static str {
            "encyclopedia"
        }
    }

    impl Needs<SessionRef> for Wiki {
        fn supply(&mut self, _session: &SessionRef) -> Result<()> {
            Ok(())
        }
    }
}

pub mod actions {
    use engine::HasWellKnownEntities;

    use super::model::*;
    use crate::library::actions::*;

    fn lookup_page_name(
        session: &SessionRef,
        world: &Entry,
        page_name: Option<&str>,
    ) -> Result<Option<Entry>, DomainError> {
        let Some(cyclo) = world.get_encyclopedia()? else {
                return Ok(None);
            };

        let cyclo = session.entry(&LookupBy::Key(&cyclo))?;
        let cyclo = cyclo.expect("TODO Dangling entity");

        if let Some(page_name) = page_name {
            let found = cyclo.get_well_known_by_name(page_name)?;
            if let Some(found) = found {
                Ok(session.entry(&LookupBy::Key(&found))?)
            } else {
                todo!("Create new pages")
            }
        } else {
            Ok(Some(cyclo))
        }
    }

    #[action]
    pub struct ReadHelpAction {
        pub page_name: Option<String>,
    }

    impl Action for ReadHelpAction {
        fn is_read_only() -> bool {
            true
        }

        fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
            let (world, _, _) = surroundings.unpack();

            let page = lookup_page_name(
                &session,
                &world,
                self.page_name.as_ref().map(|s| s.as_str()),
            )?;
            let Some(page) = page else {
                return Ok(Effect::Reply(EffectReply::Instance(Rc::new(SimpleReply::NotFound))));
            };

            let wiki = page.scope::<Wiki>()?;
            let reply: MarkdownReply = wiki.get_default().unwrap_or_else(|| "".to_owned()).into();
            Ok(Effect::Reply(EffectReply::Instance(Rc::new(reply))))
        }
    }

    #[action]
    pub struct HelpWithAction {
        pub page_name: String,
    }

    impl Action for HelpWithAction {
        fn is_read_only() -> bool {
            true
        }

        fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
            info!("editing {:?}", self.page_name);

            let (world, _, _) = surroundings.unpack();
            let page = lookup_page_name(&session, &world, Some(&self.page_name))?;
            let Some(page) = page else {
                return Ok(Effect::Reply(EffectReply::Instance(Rc::new(SimpleReply::NotFound))));
            };

            let wiki = page.scope::<Wiki>()?;
            let body: String = wiki.get_default().unwrap_or_else(|| "".to_owned()).into();
            let reply = EditorReply::new(page.key().to_string(), WorkingCopy::Markdown(body));
            Ok(Effect::Reply(EffectReply::Instance(Rc::new(reply))))
        }
    }

    #[action]
    pub struct SaveHelpAction {
        pub key: EntityKey,
        pub copy: WorkingCopy,
    }

    impl Action for SaveHelpAction {
        fn is_read_only() -> bool {
            false
        }

        fn perform(&self, session: SessionRef, _surroundings: &Surroundings) -> ReplyResult {
            info!("saving {:?}", self.key);

            match session.entry(&LookupBy::Key(&self.key))? {
                Some(entry) => {
                    match &self.copy {
                        WorkingCopy::Markdown(markdown) => {
                            let mut wiki = entry.scope_mut::<Wiki>()?;
                            wiki.set_default(markdown);
                            wiki.save()?;
                        }
                        _ => unimplemented!(),
                    }

                    Ok(SimpleReply::Done.into())
                }
                None => Ok(SimpleReply::NotFound.into()),
            }
        }
    }
}

pub mod parser {
    use super::actions::*;
    use crate::library::parser::*;

    pub struct ReadHelpParser {}

    impl ParsesActions for ReadHelpParser {
        fn try_parse_action(&self, i: &str) -> EvaluationResult {
            let (_, action) = map(alt((tag("help"), tag("wtf"))), |_s| {
                Box::new(ReadHelpAction { page_name: None }) as Box<dyn Action>
            })(i)?;

            Ok(Some(action))
        }
    }

    pub struct HelpWithParser {}

    impl ParsesActions for HelpWithParser {
        fn try_parse_action(&self, i: &str) -> EvaluationResult {
            let (_, action) = map(
                separated_pair(
                    separated_pair(tag("help"), spaces, tag("with")),
                    spaces,
                    text_to_end_of_line,
                ),
                |(_, page_name)| {
                    Box::new(HelpWithAction {
                        page_name: page_name.to_owned(),
                    }) as Box<dyn Action>
                },
            )(i)?;

            Ok(Some(action))
        }
    }
}

#[cfg(test)]
mod tests {
    use super::parser::*;
    use crate::library::tests::*;

    #[test]
    fn it_reads_default_help_on_help() -> Result<()> {
        let (_surroundings, effect) = parse_and_perform(ReadHelpParser {}, "help")?;

        assert!(matches!(effect, Effect::Ok));

        Ok(())
    }

    #[test]
    fn it_reads_default_help_on_wtf() -> Result<()> {
        let (_surroundings, effect) = parse_and_perform(ReadHelpParser {}, "wtf")?;

        assert!(matches!(effect, Effect::Ok));

        Ok(())
    }

    #[test]
    fn it_allows_helping_with_help() -> Result<()> {
        let (_surroundings, effect) = parse_and_perform(HelpWithParser {}, "help with Food")?;

        assert!(matches!(effect, Effect::Ok));

        Ok(())
    }

    /*
    #[test]
    fn it_requires_help_with_page() -> Result<()> {
        let err = parse_and_perform(HelpWithParser {}, "help with");

        assert_eq!(err.unwrap_err(), EvaluationError::ParseFailed);

        Ok(())
    }
    */
}
