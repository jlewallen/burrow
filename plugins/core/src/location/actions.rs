use crate::library::actions::*;

use super::Location;

#[action]
pub struct MoveAction {
    pub item: Item,
    pub destination: Item,
}

impl Action for MoveAction {
    fn is_read_only() -> bool
    where
        Self: Sized,
    {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        match session.find_item(surroundings, &self.item)? {
            Some(item) => match session.find_item(surroundings, &self.destination)? {
                Some(destination) => {
                    match Location::get(&item)? {
                        Some(location) => {
                            tools::move_between(&location.to_entity()?, &destination, &item)?
                        }
                        None => tools::start_carrying(&destination, &item)?,
                    };

                    Ok(SimpleReply::Done.try_into()?)
                }
                None => Ok(SimpleReply::NotFound.try_into()?),
            },
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}
