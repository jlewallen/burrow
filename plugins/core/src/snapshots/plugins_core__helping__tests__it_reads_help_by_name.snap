---
source: plugins/core/src/helping.rs
expression: effect.to_debug_json()?
---
{
  "simpleReply": "notFound"
}
